snorkel/models/context.py

```diff
@@ -158,7 +159,193 @@ def _asdict(self):
+        'polymorphic_identity': 'table',
+    }
+
+    __table_args__ = (
+        UniqueConstraint(document_id, position),
+    )
+
+    def __repr__(self):
+        return "Table(Doc: %s, Position: %s)" % (self.document.name, self.position)
+
+
+class Row(Context):
+    """A row Context in a Document."""
+    __tablename__ = 'row'
+    id = Column(Integer, ForeignKey('context.id'), primary_key=True)
+    document_id = Column(Integer, ForeignKey('document.id'))
+    table_id = Column(Integer, ForeignKey('table.id'))
+    document = relationship('Document', backref=backref('rows', cascade='all, delete-orphan'), foreign_keys=document_id)
+    table = relationship('Table', backref=backref('rows', cascade='all, delete-orphan'), foreign_keys=table_id)
+    position = Column(Integer, nullable=False)
+
+    __mapper_args__ = {
+        'polymorphic_identity': 'row',
+    }
+
+    __table_args__ = (
+        UniqueConstraint(document_id, table_id, position),
+    )
+
+    def __repr__(self):
+        return "Row(Doc: %s, Table: %s, Position: %s)" % (self.document.name, self.table.position, self.position)
+
+
+class Col(Context):
+    """A column Context in a Document."""
+    __tablename__ = 'col'
+    id = Column(Integer, ForeignKey('context.id'), primary_key=True)
+    document_id = Column(Integer, ForeignKey('document.id'))
+    table_id = Column(Integer, ForeignKey('table.id'))
+    document = relationship('Document', backref=backref('cols', cascade='all, delete-orphan'), foreign_keys=document_id)
+    table = relationship('Table', backref=backref('cols', cascade='all, delete-orphan'), foreign_keys=table_id)
+    position = Column(Integer, nullable=False)
+
+    __mapper_args__ = {
+        'polymorphic_identity': 'col',
+    }
+
+    __table_args__ = (
+        UniqueConstraint(document_id, table_id, position),
+    )
+
+    def __repr__(self):
+        return "Col(Doc: %s, Table: %s, Position: %s)" % (self.document.name, self.table.position, self.position)
+
+
+class Cell(Context):
+    """A cell Context in a Document."""
+    __tablename__ = 'cell'
+    id = Column(Integer, ForeignKey('context.id'), primary_key=True)
+    document_id = Column(Integer, ForeignKey('document.id'))
+    table_id = Column(Integer, ForeignKey('table.id'))
+    row_id =  Column(Integer, ForeignKey('row.id'))
+    col_id =  Column(Integer, ForeignKey('col.id'))
+    document = relationship('Document', backref=backref('cells', cascade='all, delete-orphan'), foreign_keys=document_id)
+    table = relationship('Table', backref=backref('cells', cascade='all, delete-orphan'), foreign_keys=table_id)
+    row = relationship('Row', backref=backref('cells', cascade='all, delete-orphan'), foreign_keys=row_id)
+    col = relationship('Col', backref=backref('cells', cascade='all, delete-orphan'), foreign_keys=col_id)
+    row_num = Column(Integer)
+    col_num = Column(Integer)
+    text = Column(Text, nullable=False)
+    html_tag = Column(Text)
+    if snorkel_postgres:
```

Comment:
Add comment here about our discussion (ideally we'd store this type of tag information hierarchically, but we store it all here for more efficient access by the feature extractor since the user / LFs probably won't touch this...)